proposals/NNNN-objc-implementation.md

```diff
@@ -0,0 +1,323 @@
+
+`@objc(CustomName) extension` also has the effect of `@objc extension`, namely, making members of the extension `@objc` by default.
+
+### `@implementation`
+
+The compiler will accept a new attribute, `@implementation`, which turns a declaration that would normally be exported to another language into a declaration that implements something imported from another language. This attribute takes no arguments; it should be used alongside another attribute specifying the foreign language, such as `@objc`.
+
+In general, `@implementation` will cause the affected declarations to be matched against imported declarations, and will emit errors if a good enough match cannot be found. It cannot necessarily be applied to all declarations that can be exported to a foreign language; for instance, `@objc @implementation` is only supported on whole extensions, not on individual members.
+
+### `@objc @implementation extension`s
+
+`@objc @implementation extension` causes an extension to be used as the implementation of an Objective-C `@interface` declaration for the class it extends. If the `@objc` attribute specifies a custom category name, it will implement a category with that name; otherwise it will implement the main `@interface` for the class (plus any class extensions).
+
+```swift
+@objc @implementation extension SomeClass {
+    // Equivalent to `@implementation SomeClass`;
+    // implements everything in `@interface SomeClass` and
+    // all `@interface SomeClass ()` extensions.
+}
+
+@objc(SomeCategory) @implementation extension SomeClass {
+    // Equivalent to `@implementation SomeClass (SomeCategory)`;
+    // implements everything in `@interface SomeClass (SomeCategory)`.
+}
+```
+
+As in any `@objc extension`, all members are implicitly `@objc` by default, and all `@objc` members are implicitly `dynamic`; unlike other `@objc extension`s, adding the `final` keyword makes the declaration *not* `@objc` by default.
+
+As a special exception to the usual rule, an `@objc @implementation extension` which implements the main Objective-C interface of a class can declare stored properties, designated and required `init`s, and `deinit`s.
+
+#### Rules
+
+An `@objc @implementation extension` must:
+
+* Extend a non-root class imported from Objective-C which does not use lightweight generics.
+* If a category name is present, match a category by that name for that class (if no category name is present, the extension matches the main interface).
+* Not declare conformances. (Conformances should be declared in the header if they are for Objective-C protocols, or in an ordinary extension otherwise.)
```

Comment:
W/r/t the rules for which members can appear in an `@objc @implementation extension` I think it might be good to contrast why we don't adopt a similar philosophy here and say that anything which isn't part of the implemented interface should be in a regular extension

Reply:
The explanation for this one is a little involved, so I'm adding an entry to Alternatives Considered.